File: rust/examples/l2cap_bridge.rs
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust version of the Python `l2cap_bridge.py` found under the `apps` folder.

use anyhow::anyhow;
use bumble::wrapper::{
    device::Device,
    l2cap::LeConnectionOrientedChannel,
    logging::{bumble_env_logging_level, py_logging_basic_config},
    transport::Transport,
};
use clap::Parser as _;
use owo_colors::OwoColorize;
use pyo3::{PyObject, PyResult, Python};
use std::{future::Future, path, sync::Arc};
use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    net::tcp::{OwnedReadHalf, OwnedWriteHalf},
    sync::{mpsc::Receiver, Mutex},
};

#[pyo3_asyncio::tokio::main]
async fn main() -> PyResult<()> {
    env_logger::builder()
        .filter_level(log::LevelFilter::Info)
        .init();

    py_logging_basic_config(bumble_env_logging_level("WARNING"))?;

    let cli = Cli::parse();

    println!("<<< connecting to HCI...");
    let transport = Transport::open(cli.hci_transport).await?;
    println!("<<< connected");

    let mut device = Device::from_config_file_with_hci(
        &cli.device_config,
        transport.source()?,
        transport.sink()?,
    )?;

    device.power_on().await?;

    match cli.subcommand {
        Subcommand::Server { tcp_host, tcp_port } => {
            let args = server_bridge::Args {
                psm: cli.psm,
                max_credits: cli.l2cap_coc_max_credits,
                mtu: cli.l2cap_coc_mtu,
                mps: cli.l2cap_coc_mps,
                tcp_host,
                tcp_port,
            };

            server_bridge::start(&args, &mut device).await?
        }
        Subcommand::Client {
            bluetooth_address,
            tcp_host,
            tcp_port,
        } => {
            let args = client_bridge::Args {
                psm: cli.psm,
                max_credits: cli.l2cap_coc_max_credits,
                mtu: cli.l2cap_coc_mtu,
                mps: cli.l2cap_coc_mps,
                bluetooth_address,
                tcp_host,
                tcp_port,
            };

            client_bridge::start(&args, &mut device).await?
        }
    };

    // wait until user kills the process
    tokio::signal::ctrl_c().await?;

    Ok(())
}

/// L2CAP CoC server bridge: waits for a peer to connect an L2CAP CoC channel
/// on a specified PSM. When the connection is made, the bridge connects a TCP
/// socket to a remote host and bridges the data in both directions, with flow
/// control.
/// When the L2CAP CoC channel is closed, the bridge disconnects the TCP socket
/// and waits for a new L2CAP CoC channel to be connected.
/// When the TCP connection is closed by the TCP server, XXXX
mod server_bridge {
    use crate::{
        proxy_l2cap_rx_to_tcp_tx, proxy_tcp_rx_to_l2cap_tx, run_future_with_current_task_locals,
        BridgeData,
    };
    use bumble::wrapper::{device::Device, hci::HciConstant, l2cap::LeConnectionOrientedChannel};
    use futures::executor::block_on;
    use owo_colors::OwoColorize;
    use pyo3::{PyResult, Python};
    use std::{sync::Arc, time::Duration};
    use tokio::{
        join,
        net::TcpStream,
        select,
        sync::{mpsc, Mutex},
    };

    pub struct Args {
        pub psm: u16,
        pub max_credits: u16,
        pub mtu: u16,
        pub mps: u16,
        pub tcp_host: String,
        pub tcp_port: u16,
    }

    pub async fn start(args: &Args, device: &mut Device) -> PyResult<()> {
        let host = args.tcp_host.clone();
        let port = args.tcp_port;
        device.register_l2cap_channel_server(
            args.psm,
            move |_py, l2cap_channel| {
                let channel_info = match l2cap_channel.debug_string() {
                    Ok(info_string) => info_string,
                    Err(py_err) => format!("failed to get l2cap channel info ({})", py_err),
                };
                println!("{} {channel_info}", "*** L2CAP channel:".cyan());

                let host = host.clone();
                // Ensure Python event loop is available to l2cap `disconnect`
                let _ = run_future_with_current_task_locals(handle_connection_oriented_channel(
                    l2cap_channel,
                    host,
                    port,
                ));
                Ok(())
            },
            Some(args.max_credits),
            Some(args.mtu),
            Some(args.mps),
        )?;

        println!(
            "{}",
            format!("### Listening for CoC connection on PSM {}", args.psm).yellow()
        );

        device.on_connection(|_py, mut connection| {
            let connection_info = match connection.debug_string() {
                Ok(info_string) => info_string,
                Err(py_err) => format!("failed to get connection info ({})", py_err),
            };
            println!(
                "{} {}",
                "@@@ Bluetooth connection: ".green(),
                connection_info,
            );
            connection.on_disconnection(|_py, reason| {
                let disconnection_info = match HciConstant::error_name(reason) {
                    Ok(info_string) => info_string,
                    Err(py_err) => format!("failed to get disconnection error name ({})", py_err),
                };
                println!(
                    "{} {}",
                    "@@@ Bluetooth disconnection: ".red(),
                    disconnection_info,
                );
                Ok(())
            })?;
            Ok(())
        })?;

        device.start_advertising(false).await?;

        Ok(())
    }

    async fn handle_connection_oriented_channel(
        mut l2cap_channel: LeConnectionOrientedChannel,
        tcp_host: String,
        tcp_port: u16,
    ) -> PyResult<()> {
        let (l2cap_to_tcp_tx, mut l2cap_to_tcp_rx) = mpsc::channel::<BridgeData>(10);

        // Set callback (`set_sink`) for when l2cap data is received.
        let l2cap_to_tcp_tx_clone = l2cap_to_tcp_tx.clone();
        l2cap_channel
            .set_sink(move |_py, sdu| {
                block_on(l2cap_to_tcp_tx_clone.send(BridgeData::Data(sdu.into())))
                    .expect("failed to channel data to tcp");
                Ok(())
            })
            .expect("failed to set sink for l2cap connection");

        // Set l2cap callback for when the channel is closed.
        l2cap_channel
            .on_close(move |_py| {
                println!("{}", "*** L2CAP channel closed".red());
                block_on(l2cap_to_tcp_tx.send(BridgeData::CloseSignal))
                    .expect("failed to channel close signal to tcp");
                Ok(())
            })
            .expect("failed to set on_close callback for l2cap channel");

        println!(
            "{}",
            format!("### Connecting to TCP {tcp_host}:{tcp_port}...").yellow()
        );

        let l2cap_channel = Arc::new(Mutex::new(Some(l2cap_channel)));
        let tcp_stream = match TcpStream::connect(format!("{tcp_host}:{tcp_port}")).await {
            Ok(stream) => {
                println!("{}", "### Connected".green());
                Some(stream)
            }
            Err(err) => {
                println!("{}", format!("!!! Connection failed: {err}").red());
                if let Some(channel) = l2cap_channel.lock().await.take() {
                    // Bumble might enter an invalid state if disconnection request is received from
                    // l2cap client before receiving a disconnection response from the same client,
                    // blocking this async call from returning.
                    // See: https://github.com/google/bumble/issues/257
                    select! {
                        res = channel.disconnect() => {
                            let _ = res.map_err(|e| eprintln!("Failed to call disconnect on l2cap channel: {e}"));
                        },
                        _ = tokio::time::sleep(Duration::from_secs(1)) => eprintln!("Timed out while calling disconnect on l2cap channel."),
                    }
                }
                None
            }
        };

        match tcp_stream {
            None => {
                while let Some(bridge_data) = l2cap_to_tcp_rx.recv().await {
                    match bridge_data {
                        BridgeData::Data(sdu) => {
                            println!("{}", format!("<<< [L2CAP SDU]: {} bytes", sdu.len()).cyan());
                            println!("{}", "!!! TCP socket not open, dropping".red())
                        }
                        BridgeData::CloseSignal => break,
                    }
                }
            }
            Some(tcp_stream) => {
                let (tcp_reader, tcp_writer) = tcp_stream.into_split();

                // Do tcp stuff when something happens on the l2cap channel.
                let handle_l2cap_data_future =
                    proxy_l2cap_rx_to_tcp_tx(l2cap_to_tcp_rx, tcp_writer, l2cap_channel.clone());

                // Do l2cap stuff when something happens on tcp.
                let handle_tcp_data_future =
                    proxy_tcp_rx_to_l2cap_tx(tcp_reader, l2cap_channel.clone(), false);

                let (handle_l2cap_result, handle_tcp_result) =
                    join!(handle_l2cap_data_future, handle_tcp_data_future);

                if let Err(e) = handle_l2cap_result {
                    println!("!!! Error: {e}");
                }

                if let Err(e) = handle_tcp_result {
                    println!("!!! Error: {e}");
                }
            }
        };

        Python::with_gil(|_| {
            // Must hold GIL at least once while/after dropping for Python heap object to ensure
            // de-allocation.
            drop(l2cap_channel);
        });

        Ok(())
    }
}

/// L2CAP CoC client bridge: connects to a BLE device, then waits for an inbound
/// TCP connection on a specified port number. When a TCP client connects, an
/// L2CAP CoC channel connection to the BLE device is established, and the data
/// is bridged in both directions, with flow control.
/// When the TCP connection is closed by the client, the L2CAP CoC channel is
/// disconnected, but the connection to the BLE device remains, ready for a new
/// TCP client to connect.
/// When the L2CAP CoC channel is closed, XXXX
mod client_bridge {
    use crate::{
        proxy_l2cap_rx_to_tcp_tx, proxy_tcp_rx_to_l2cap_tx, run_future_with_current_task_locals,
        BridgeData,
    };
    use bumble::wrapper::{
        device::{Connection, Device},
        hci::HciConstant,
    };
    use futures::executor::block_on;
    use owo_colors::OwoColorize;
    use pyo3::{PyResult, Python};
    use std::{net::SocketAddr, sync::Arc};
    use tokio::{
        join,
        net::{TcpListener, TcpStream},
        sync::{mpsc, Mutex},
    };

    pub struct Args {
        pub psm: u16,
        pub max_credits: u16,
        pub mtu: u16,
        pub mps: u16,
        pub bluetooth_address: String,
        pub tcp_host: String,
        pub tcp_port: u16,
    }

    pub async fn start(args: &Args, device: &mut Device) -> PyResult<()> {
        println!(
            "{}",
            format!("### Connecting to {}...", args.bluetooth_address).yellow()
        );
        let mut ble_connection = device.connect(&args.bluetooth_address).await?;
        ble_connection.on_disconnection(|_py, reason| {
            let disconnection_info = match HciConstant::error_name(reason) {
                Ok(info_string) => info_string,
                Err(py_err) => format!("failed to get disconnection error name ({})", py_err),
            };
            println!(
                "{} {}",
                "@@@ Bluetooth disconnection: ".red(),
                disconnection_info,
            );
            Ok(())
        })?;

        // Start the TCP server.
        let listener = TcpListener::bind(format!("{}:{}", args.tcp_host, args.tcp_port))
            .await
            .expect("failed to bind tcp to address");
        println!(
            "{}",
            format!(
                "### Listening for TCP connections on port {}",
                args.tcp_port
            )
            .magenta()
        );

        let psm = args.psm;
        let max_credits = args.max_credits;
        let mtu = args.mtu;
        let mps = args.mps;
        let ble_connection = Arc::new(ble_connection);
        // Ensure Python event loop is available to l2cap `disconnect`
        let _ = run_future_with_current_task_locals(async move {
            while let Ok((tcp_stream, addr)) = listener.accept().await {
                let ble_connection = ble_connection.clone();
                let _ = run_future_with_current_task_locals(handle_tcp_connection(
                    ble_connection,
                    tcp_stream,
                    addr,
                    psm,
                    max_credits,
                    mtu,
                    mps,
                ));
            }
            Ok(())
        });
        Ok(())
    }

    async fn handle_tcp_connection(
        ble_connection: Arc<Connection>,
        tcp_stream: TcpStream,
        addr: SocketAddr,
        psm: u16,
        max_credits: u16,
        mtu: u16,
        mps: u16,
    ) -> PyResult<()> {
        println!("{}", format!("<<< TCP connection from {}", addr).magenta());
        println!(
            "{}",
            format!(">>> Opening L2CAP channel on PSM = {}", psm).yellow()
        );

        let mut l2cap_channel = match ble_connection
            .open_l2cap_channel(psm, Some(max_credits), Some(mtu), Some(mps))
            .await
        {
            Ok(channel) => channel,
            Err(e) => {
                println!("{}", format!("!!! Connection failed: {e}").red());
                // TCP stream will get dropped after returning, automatically shutting it down.
                return Err(e);
            }
        };
        let channel_info = match l2cap_channel.debug_string() {
            Ok(info_string) => info_string,
            Err(py_err) => format!("failed to get l2cap channel info ({})", py_err),
        };

        println!("{}{}", "*** L2CAP channel: ".cyan(), channel_info);

        let (l2cap_to_tcp_tx, l2cap_to_tcp_rx) = mpsc::channel::<BridgeData>(10);

        // Set l2cap callback (`set_sink`) for when data is received.
        let l2cap_to_tcp_tx_clone = l2cap_to_tcp_tx.clone();
        l2cap_channel
            .set_sink(move |_py, sdu| {
                block_on(l2cap_to_tcp_tx_clone.send(BridgeData::Data(sdu.into())))
                    .expect("failed to channel data to tcp");
                Ok(())
            })
            .expect("failed to set sink for l2cap connection");

        // Set l2cap callback for when the channel is closed.
        l2cap_channel
            .on_close(move |_py| {
                println!("{}", "*** L2CAP channel closed".red());
                block_on(l2cap_to_tcp_tx.send(BridgeData::CloseSignal))
                    .expect("failed to channel close signal to tcp");
                Ok(())
            })
            .expect("failed to set on_close callback for l2cap channel");

        let l2cap_channel = Arc::new(Mutex::new(Some(l2cap_channel)));
        let (tcp_reader, tcp_writer) = tcp_stream.into_split();

        // Do tcp stuff when something happens on the l2cap channel.
        let handle_l2cap_data_future =
            proxy_l2cap_rx_to_tcp_tx(l2cap_to_tcp_rx, tcp_writer, l2cap_channel.clone());

        // Do l2cap stuff when something happens on tcp.
        let handle_tcp_data_future =
            proxy_tcp_rx_to_l2cap_tx(tcp_reader, l2cap_channel.clone(), true);

        let (handle_l2cap_result, handle_tcp_result) =
            join!(handle_l2cap_data_future, handle_tcp_data_future);

        if let Err(e) = handle_l2cap_result {
            println!("!!! Error: {e}");
        }

        if let Err(e) = handle_tcp_result {
            println!("!!! Error: {e}");
        }

        Python::with_gil(|_| {
            // Must hold GIL at least once while/after dropping for Python heap object to ensure
            // de-allocation.
            drop(l2cap_channel);
        });

        Ok(())
    }
}

/// Used for channeling data from Python callbacks to a Rust consumer.
enum BridgeData {
    Data(Vec<u8>),
    CloseSignal,
}

async fn proxy_l2cap_rx_to_tcp_tx(
    mut l2cap_data_receiver: Receiver<BridgeData>,
    mut tcp_writer: OwnedWriteHalf,
    l2cap_channel: Arc<Mutex<Option<LeConnectionOrientedChannel>>>,
) -> anyhow::Result<()> {
    while let Some(bridge_data) = l2cap_data_receiver.recv().await {
        match bridge_data {
            BridgeData::Data(sdu) => {
                println!("{}", format!("<<< [L2CAP SDU]: {} bytes", sdu.len()).cyan());
                tcp_writer
                    .write_all(sdu.as_ref())
                    .await
                    .map_err(|_| anyhow!("Failed to write to tcp stream"))?;
                tcp_writer
                    .flush()
                    .await
                    .map_err(|_| anyhow!("Failed to flush tcp stream"))?;
            }
            BridgeData::CloseSignal => {
                l2cap_channel.lock().await.take();
                tcp_writer
                    .shutdown()
                    .await
                    .map_err(|_| anyhow!("Failed to shut down write half of tcp stream"))?;
                return Ok(());
            }
        }
    }
    Ok(())
}

async fn proxy_tcp_rx_to_l2cap_tx(
    mut tcp_reader: OwnedReadHalf,
    l2cap_channel: Arc<Mutex<Option<LeConnectionOrientedChannel>>>,
    drain_l2cap_after_write: bool,
) -> PyResult<()> {
    let mut buf = [0; 4096];
    loop {
        match tcp_reader.read(&mut buf).await {
            Ok(len) => {
                if len == 0 {
                    println!("{}", "!!! End of stream".yellow());

                    if let Some(channel) = l2cap_channel.lock().await.take() {
                        channel.disconnect().await.map_err(|e| {
                            eprintln!("Failed to call disconnect on l2cap channel: {e}");
                            e
                        })?;
                    }
                    return Ok(());
                }

                println!("{}", format!("<<< [TCP DATA]: {len} bytes").blue());
                match l2cap_channel.lock().await.as_mut() {
                    None => {
                        println!("{}", "!!! L2CAP channel not connected, dropping".red());
                        return Ok(());
                    }
                    Some(channel) => {
                        channel.write(&buf[..len])?;
                        if drain_l2cap_after_write {
                            channel.drain().await?;
                        }
                    }
                }
            }
            Err(e) => {
                println!("{}", format!("!!! TCP connection lost: {}", e).red());
                if let Some(channel) = l2cap_channel.lock().await.take() {
                    let _ = channel.disconnect().await.map_err(|e| {
                        eprintln!("Failed to call disconnect on l2cap channel: {e}");
                    });
                }
                return Err(e.into());
            }
        }
    }
}

/// Copies the current thread's task locals into a Python "awaitable" and encapsulates it in a Rust
/// future, running it as a Python Task.
///
/// If the calling thread has a Python event loop, then the Python Task will too.
pub fn run_future_with_current_task_locals<F>(
    fut: F,
) -> PyResult<impl Future<Output = PyResult<PyObject>> + Send>
where
    F: Future<Output = PyResult<()>> + Send + 'static,
{
    Python::with_gil(|py| {
        let locals = pyo3_asyncio::tokio::get_current_locals(py)?;
        let future = pyo3_asyncio::tokio::scope(locals.clone(), fut);
        pyo3_asyncio::tokio::future_into_py_with_locals(py, locals.clone(), future)
            .and_then(pyo3_asyncio::tokio::into_future)
    })
}

#[derive(clap::Parser)]
#[command(author, version, about, long_about = None)]
struct Cli {
    #[command(subcommand)]
    pub(crate) subcommand: Subcommand,

    /// Device configuration file.
    ///
    /// See, for instance, `examples/device1.json` in the Python project.
    #[arg(long)]
    device_config: path::PathBuf,
    /// Bumble transport spec.
    ///
    /// <https://google.github.io/bumble/transports/index.html>
    #[arg(long)]
    hci_transport: String,

    /// PSM for L2CAP Connection-oriented Channel.
    ///
    /// Must be in the range [0, 65535].
    #[arg(long, default_value_t = 1234)]
    psm: u16,

    /// Maximum L2CAP CoC Credits.
    ///
    /// Must be in the range [1, 65535].
    #[arg(long, default_value_t = 128, value_parser = clap::value_parser!(u16).range(1..))]
    l2cap_coc_max_credits: u16,

    /// L2CAP CoC MTU
    ///
    /// Must be in the range [23, 65535].
    #[arg(long, default_value_t = 1022, value_parser = clap::value_parser!(u16).range(23..))]
    l2cap_coc_mtu: u16,

    /// L2CAP CoC MPS
    ///
    /// Must be in the range [23, 65535].
    #[arg(long, default_value_t = 1024, value_parser = clap::value_parser!(u16).range(23..))]
    l2cap_coc_mps: u16,
}

#[derive(clap::Subcommand)]
enum Subcommand {
    /// Starts an L2CAP server
    Server {
        /// TCP host that the l2cap server will connect to.
        /// Data is bridged like so:
        ///     TCP server <-> (TCP client / **L2CAP server**) <-> (L2CAP client / TCP server) <-> TCP client
        #[arg(long, default_value = "localhost")]
        tcp_host: String,
        /// TCP port that the server will connect to.
        ///
        /// Must be in the range [1, 65535].
        #[arg(long, default_value_t = 9544)]
        tcp_port: u16,
    },
    /// Starts an L2CAP client
    Client {
        /// L2cap server address that this l2cap client will connect to.
        bluetooth_address: String,
        /// TCP host that the l2cap client will bind to and listen for incoming TCP connections.
        /// Data is bridged like so:
        ///     TCP client <-> (TCP server / **L2CAP client**) <-> (L2CAP server / TCP client) <-> TCP Client
        #[arg(long, default_value = "localhost")]
        tcp_host: String,
        /// TCP port that the client will connect to.
        ///
        /// Must be in the range [1, 65535].
        #[arg(long, default_value_t = 9543)]
        tcp_port: u16,
    },
}

File: apps/l2cap_bridge.py
# Copyright 2021-2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -----------------------------------------------------------------------------
# Imports
# -----------------------------------------------------------------------------
import asyncio
import logging
import os
import click

from bumble import l2cap
from bumble.colors import color
from bumble.transport import open_transport_or_link
from bumble.device import Device
from bumble.utils import FlowControlAsyncPipe
from bumble.hci import HCI_Constant


# -----------------------------------------------------------------------------
class ServerBridge:
    """
    L2CAP CoC server bridge: waits for a peer to connect an L2CAP CoC channel
    on a specified PSM. When the connection is made, the bridge connects a TCP
    socket to a remote host and bridges the data in both directions, with flow
    control.
    When the L2CAP CoC channel is closed, the bridge disconnects the TCP socket
    and waits for a new L2CAP CoC channel to be connected.
    When the TCP connection is closed by the TCP server, XXXX
    """

    def __init__(self, psm, max_credits, mtu, mps, tcp_host, tcp_port):
        self.psm = psm
        self.max_credits = max_credits
        self.mtu = mtu
        self.mps = mps
        self.tcp_host = tcp_host
        self.tcp_port = tcp_port

    async def start(self, device: Device) -> None:
        # Listen for incoming L2CAP CoC connections
        device.create_l2cap_server(
            spec=l2cap.LeCreditBasedChannelSpec(
                psm=self.psm, mtu=self.mtu, mps=self.mps, max_credits=self.max_credits
            ),
            handler=self.on_coc,
        )
        print(color(f'### Listening for CoC connection on PSM {self.psm}', 'yellow'))

        def on_ble_connection(connection):
            def on_ble_disconnection(reason):
                print(
                    color('@@@ Bluetooth disconnection:', 'red'),
                    HCI_Constant.error_name(reason),
                )

            print(color('@@@ Bluetooth connection:', 'green'), connection)
            connection.on('disconnection', on_ble_disconnection)

        device.on('connection', on_ble_connection)

        await device.start_advertising(auto_restart=True)

    # Called when a new L2CAP connection is established
    def on_coc(self, l2cap_channel):
        print(color('*** L2CAP channel:', 'cyan'), l2cap_channel)

        class Pipe:
            def __init__(self, bridge, l2cap_channel):
                self.bridge = bridge
                self.tcp_transport = None
                self.l2cap_channel = l2cap_channel

                l2cap_channel.on('close', self.on_l2cap_close)
                l2cap_channel.sink = self.on_coc_sdu

            async def connect_to_tcp(self):
                # Connect to the TCP server
                print(
                    color(
                        f'### Connecting to TCP {self.bridge.tcp_host}:'
                        f'{self.bridge.tcp_port}...',
                        'yellow',
                    )
                )

                class TcpClientProtocol(asyncio.Protocol):
                    def __init__(self, pipe):
                        self.pipe = pipe

                    def connection_lost(self, exc):
                        print(color(f'!!! TCP connection lost: {exc}', 'red'))
                        if self.pipe.l2cap_channel is not None:
                            asyncio.create_task(self.pipe.l2cap_channel.disconnect())

                    def data_received(self, data):
                        print(color(f'<<< [TCP DATA]: {len(data)} bytes', 'blue'))
                        self.pipe.l2cap_channel.write(data)

                try:
                    (
                        self.tcp_transport,
                        _,
                    ) = await asyncio.get_running_loop().create_connection(
                        lambda: TcpClientProtocol(self),
                        host=self.bridge.tcp_host,
                        port=self.bridge.tcp_port,
                    )
                    print(color('### Connected', 'green'))
                except Exception as error:
                    print(color(f'!!! Connection failed: {error}', 'red'))
                    await self.l2cap_channel.disconnect()

            def on_l2cap_close(self):
                print(color('*** L2CAP channel closed', 'red'))
                self.l2cap_channel = None
                if self.tcp_transport is not None:
                    self.tcp_transport.close()

            def on_coc_sdu(self, sdu):
                print(color(f'<<< [L2CAP SDU]: {len(sdu)} bytes', 'cyan'))
                if self.tcp_transport is None:
                    print(color('!!! TCP socket not open, dropping', 'red'))
                    return
                self.tcp_transport.write(sdu)

        pipe = Pipe(self, l2cap_channel)

        asyncio.create_task(pipe.connect_to_tcp())


# -----------------------------------------------------------------------------
class ClientBridge:
    """
    L2CAP CoC client bridge: connects to a BLE device, then waits for an inbound
    TCP connection on a specified port number. When a TCP client connects, an
    L2CAP CoC channel connection to the BLE device is established, and the data
    is bridged in both directions, with flow control.
    When the TCP connection is closed by the client, the L2CAP CoC channel is
    disconnected, but the connection to the BLE device remains, ready for a new
    TCP client to connect.
    When the L2CAP CoC channel is closed, XXXX
    """

    READ_CHUNK_SIZE = 4096

    def __init__(self, psm, max_credits, mtu, mps, address, tcp_host, tcp_port):
        self.psm = psm
        self.max_credits = max_credits
        self.mtu = mtu
        self.mps = mps
        self.address = address
        self.tcp_host = tcp_host
        self.tcp_port = tcp_port

    async def start(self, device):
        print(color(f'### Connecting to {self.address}...', 'yellow'))
        connection = await device.connect(self.address)
        print(color('### Connected', 'green'))

        # Called when the BLE connection is disconnected
        def on_ble_disconnection(reason):
            print(
                color('@@@ Bluetooth disconnection:', 'red'),
                HCI_Constant.error_name(reason),
            )

        connection.on('disconnection', on_ble_disconnection)

        # Called when a TCP connection is established
        async def on_tcp_connection(reader, writer):
            peer_name = writer.get_extra_info('peer_name')
            print(color(f'<<< TCP connection from {peer_name}', 'magenta'))

            def on_coc_sdu(sdu):
                print(color(f'<<< [L2CAP SDU]: {len(sdu)} bytes', 'cyan'))
                l2cap_to_tcp_pipe.write(sdu)

            def on_l2cap_close():
                print(color('*** L2CAP channel closed', 'red'))
                l2cap_to_tcp_pipe.stop()
                writer.close()

            # Connect a new L2CAP channel
            print(color(f'>>> Opening L2CAP channel on PSM = {self.psm}', 'yellow'))
            try:
                l2cap_channel = await connection.create_l2cap_channel(
                    spec=l2cap.LeCreditBasedChannelSpec(
                        psm=self.psm,
                        max_credits=self.max_credits,
                        mtu=self.mtu,
                        mps=self.mps,
                    )
                )
                print(color('*** L2CAP channel:', 'cyan'), l2cap_channel)
            except Exception as error:
                print(color(f'!!! Connection failed: {error}', 'red'))
                writer.close()
                return

            l2cap_channel.sink = on_coc_sdu
            l2cap_channel.on('close', on_l2cap_close)

            # Start a flow control pipe from L2CAP to TCP
            l2cap_to_tcp_pipe = FlowControlAsyncPipe(
                l2cap_channel.pause_reading,
                l2cap_channel.resume_reading,
                writer.write,
                writer.drain,
            )
            l2cap_to_tcp_pipe.start()

            # Pipe data from TCP to L2CAP
            while True:
                try:
                    data = await reader.read(self.READ_CHUNK_SIZE)

                    if len(data) == 0:
                        print(color('!!! End of stream', 'red'))
                        await l2cap_channel.disconnect()
                        return

                    print(color(f'<<< [TCP DATA]: {len(data)} bytes', 'blue'))
                    l2cap_channel.write(data)
                    await l2cap_channel.drain()
                except Exception as error:
                    print(f'!!! Exception: {error}')
                    break

            writer.close()
            print(color('~~~ Bye bye', 'magenta'))

        await asyncio.start_server(
            on_tcp_connection,
            host=self.tcp_host if self.tcp_host != '_' else None,
            port=self.tcp_port,
        )
        print(
            color(
                f'### Listening for TCP connections on port {self.tcp_port}', 'magenta'
            )
        )


# -----------------------------------------------------------------------------
async def run(device_config, hci_transport, bridge):
    print('<<< connecting to HCI...')
    async with await open_transport_or_link(hci_transport) as (hci_source, hci_sink):
        print('<<< connected')

        device = Device.from_config_file_with_hci(device_config, hci_source, hci_sink)

        # Let's go
        await device.power_on()
        await bridge.start(device)

        # Wait until the transport terminates
        await hci_source.wait_for_termination()


# -----------------------------------------------------------------------------
@click.group()
@click.pass_context
@click.option('--device-config', help='Device configuration file', required=True)
@click.option('--hci-transport', help='HCI transport', required=True)
@click.option('--psm', help='PSM for L2CAP CoC', type=int, default=1234)
@click.option(
    '--l2cap-coc-max-credits',
    help='Maximum L2CAP CoC Credits',
    type=click.IntRange(1, 65535),
    default=128,
)
@click.option(
    '--l2cap-coc-mtu',
    help='L2CAP CoC MTU',
    type=click.IntRange(23, 65535),
    default=1022,
)
@click.option(
    '--l2cap-coc-mps',
    help='L2CAP CoC MPS',
    type=click.IntRange(23, 65533),
    default=1024,
)
def cli(
    context,
    device_config,
    hci_transport,
    psm,
    l2cap_coc_max_credits,
    l2cap_coc_mtu,
    l2cap_coc_mps,
):
    context.ensure_object(dict)
    context.obj['device_config'] = device_config
    context.obj['hci_transport'] = hci_transport
    context.obj['psm'] = psm
    context.obj['max_credits'] = l2cap_coc_max_credits
    context.obj['mtu'] = l2cap_coc_mtu
    context.obj['mps'] = l2cap_coc_mps


# -----------------------------------------------------------------------------
@cli.command()
@click.pass_context
@click.option('--tcp-host', help='TCP host', default='localhost')
@click.option('--tcp-port', help='TCP port', default=9544)
def server(context, tcp_host, tcp_port):
    bridge = ServerBridge(
        context.obj['psm'],
        context.obj['max_credits'],
        context.obj['mtu'],
        context.obj['mps'],
        tcp_host,
        tcp_port,
    )
    asyncio.run(run(context.obj['device_config'], context.obj['hci_transport'], bridge))


# -----------------------------------------------------------------------------
@cli.command()
@click.pass_context
@click.argument('bluetooth-address')
@click.option('--tcp-host', help='TCP host', default='_')
@click.option('--tcp-port', help='TCP port', default=9543)
def client(context, bluetooth_address, tcp_host, tcp_port):
    bridge = ClientBridge(
        context.obj['psm'],
        context.obj['max_credits'],
        context.obj['mtu'],
        context.obj['mps'],
        bluetooth_address,
        tcp_host,
        tcp_port,
    )
    asyncio.run(run(context.obj['device_config'], context.obj['hci_transport'], bridge))


# -----------------------------------------------------------------------------
logging.basicConfig(level=os.environ.get('BUMBLE_LOGLEVEL', 'WARNING').upper())
if __name__ == '__main__':
    cli(obj={})  # pylint: disable=no-value-for-parameter
